sys/net/application_layer/gcoap/dns.c

```diff
@@ -0,0 +1,837 @@
+
+    coap_block1_finish(&slicer);
+
+    if ((!IS_USED(MODULE_GCOAP_DNS_ASYNC) ||
+         !(context->method & GCOAP_DNS_ASYNC_FLAG)) &&
+        (slicer.start == 0)) {
+        mutex_lock(&context->sync.resp_wait);
+    }
+    len = gcoap_req_send((uint8_t *)pdu->hdr, len, remote, _resp_handler,
+                         context);
+    if (len <= 0) {
+        printf("gcoap_dns: msg send failed: %d\n", (int)len);
+        return len;
+    }
+    return len;
+}
+
+static ssize_t _req_post_fetch(gcoap_dns_ctx_t *context)
+{
+    coap_pkt_t *pdu = context->pkt;
+    ssize_t len;
+    uint8_t method = _decode_ctx_method(context);
+
+    if ((len = _set_remote(&_uri_comp, &_remote)) < 0) {
+        return len;
+    }
+    if (context->dns_buf_len > CONFIG_GCOAP_DNS_BLOCK_SIZE) {
+        context->cur_blk_num = 0U;
+        return _do_block_post_fetch(pdu, &_remote, context);
+    }
+    else {
+        len = _post_fetch_init(pdu, method);
+        if (len < 0) {
+            return len;
+        }
+        if (CONFIG_GCOAP_DNS_PDU_BUF_SIZE < CONFIG_DNS_MSG_LEN) {
+            coap_block1_t block;
+
+            coap_block_object_init(&block, 0, CONFIG_GCOAP_DNS_BLOCK_SIZE, 0);
+            if (coap_opt_add_block2_control(pdu, &block) < 0) {
+                DEBUG("gcoap_dns: unable to add Block2 option to %s\n",
+                      METHOD_POST_FETCH_STR(method));
+                return -ENOBUFS;
+            }
+        }
+        if (_is_proxied()) {
+            if (coap_opt_add_proxy_uri(pdu, _uri_buf) < 0) {
+                DEBUG("gcoap_dns: unable to add Proxy-URI option to %s\n",
+                      METHOD_POST_FETCH_STR(method));
+                return -ENOBUFS;
+            }
+        }
+        len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+        memcpy(pdu->payload, context->dns_buf, context->dns_buf_len);
+        if (!IS_USED(MODULE_GCOAP_DNS_ASYNC) ||
+            !(context->method & GCOAP_DNS_ASYNC_FLAG)) {
+            mutex_lock(&context->sync.resp_wait);
+        }
+        return gcoap_req_send((uint8_t *)pdu->hdr,
+                              len + context->dns_buf_len,
+                              &_remote, _resp_handler, context);
+    }
+}
+
+static void _resp_handler(const gcoap_request_memo_t *memo, coap_pkt_t *pdu,
+                          const sock_udp_ep_t *remote)
+{
+    coap_block1_t block;
+    gcoap_dns_ctx_t *context = memo->context;
+    void *data;
+    size_t data_len;
+    int family = context->family;
+
+    if (memo->state == GCOAP_MEMO_TIMEOUT) {
+        printf("gcoap_dns: CoAP request timed out\n");
+        context->res = -ETIMEDOUT;
+        goto unlock;
+    }
+    else if (memo->state == GCOAP_MEMO_ERR) {
+        printf("gcoap_dns: error in response\n");
+        context->res = -EBADMSG;
+        goto unlock;
+    }
+
+    if ((remote->port != _remote.port) ||
+        (remote->family != _remote.family) ||
+        (!ipv6_addr_is_unspecified((ipv6_addr_t *)_remote.addr.ipv6) &&
```

Comment:
One should think so, but I've got this error returned often enough that I fear it is not. Mind you, on the medium (i.e. the wireshark dump) the remote address _is_ correct, it's just that gcoap itself sometimes reports different remotes... so maybe this check _is_ a bit paranoid, but still revealing some underlying bug.

Reply:
Is there any test case you can put in a report, to have a good issue number to go both in the PR notes and into the code (so that when the issue is finally resolved, we can remove the check)?

Reply:
Not that I know of :(.